### PR TITLE
Deprecate #shared tag, use not #volatile and volatile storage key instead to register shareable stores.

### PR DESCRIPTION
**Pull request by mmandlis**
and distinguish between 'ramdisk' and 'volatile' in runtime.ts and java code

### CODE REVIEW

File: src/runtime/runtime.ts [hunk 1]
```diff
@@ -196,27 +196,17 @@ export class Runtime {
     return [...this.arcById.values()].find(arc => !!arc.activeRecipe.findParticle(particleId));
   }
 
-  // TODO: This is a temporary method to allow sharing stores with other Arcs.
+  // This is a temporary method to allow sharing stores with other Arcs.
```
Comment (mmandlis) [inline src/runtime/runtime.ts]: note to self: rework this change, if https://github.com/PolymerLabs/arcs/pull/4346 lands sooner.

File: src/runtime/arc.ts [hunk 2]
```diff
@@ -175,10 +175,6 @@ constructor({id, context, pecFactories, slotComposer, loader, storageKey, storag
     }
 
     DriverFactory.unregister(this.volatileStorageDriverProvider);
-
-    for (const store of this._stores) {
-      Runtime.getRuntime().unregisterStore(store.id, [...this.findStoreTags(store)]);
```
Comment (shaper) [inline src/runtime/arc.ts]: Is removing this ok? I worry that never unregistering stores will lead to data leaks (from a memory utilization perspective). I don't have much of the history around this code, perhaps @galganif can comment when he is in.

File: src/runtime/tests/runtime-test.ts [hunk 3]
```diff
@@ -83,23 +84,52 @@ describe('Runtime', () => {
...
+    assert.lengthOf(runtime.context.stores, 2);
+
+    ramdiskArc.dispose();
+    assert.lengthOf(runtime.context.stores, 2);
+
+    const volatileArc2 = runtime.runArc('test-arc-v1', volatileStorageKeyPrefixForTest());
+    const recipe2 = await runtime.resolveRecipe(volatileArc2, manifest.recipes[1]);
+    assert.isTrue(recipe2 && recipe2.isResolved());
+    await volatileArc2.instantiate(recipe2);
+
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply src/runtime/tests/runtime-test.ts]: i'm testing the recipe was successfully instantiated. added another assert for active recipe handle's storage key.

File: src/runtime/tests/runtime-test.ts [hunk 4]
```diff
@@ -83,23 +84,52 @@ describe('Runtime', () => {
...
+    await volatileArc1.instantiate(recipe1);
+    volatileArc1.dispose();
+
+    volatileArc.dispose();
+    assert.lengthOf(runtime.context.stores, 2);
+
+    ramdiskArc.dispose();
+    assert.lengthOf(runtime.context.stores, 2);
+
+    const volatileArc2 = runtime.runArc('test-arc-v1', volatileStorageKeyPrefixForTest());
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply src/runtime/tests/runtime-test.ts]: done, thanks.

File: src/runtime/arc.ts (hunk 2 above)
Comment (mmandlis) [reply src/runtime/arc.ts]: we don't currently distinguish between arc that is being torn down (and can be spun up again) and deleted (no longer exists).
in the former case, we still want its stores to be available to other arcs (for example, if there is an ingestion arc that is only actively running when a new data comes in and then goes down again, we want other arcs to have access to the ingested data unrelated).
the latter case isn't supported at the moment (and needs to be as part of the upcoming better defined arc's lifecycle).

**Comment by mmandlis** [comment]
@shans can you, please, take a look?

File: src/runtime/manifest.ts [hunk 5]
```diff
@@ -1364,25 +1364,15 @@ ${e.message}
 
   // TODO: This is a temporary method to allow sharing stores with other Arcs.
   registerStore(store: UnifiedStore, tags: string[]): void {
-    if (!this.findStoreById(store.id) && tags.includes('shared')) {
+    // Only register stores that have non-volatile storage key and don't have a
+    // #volatile tag.
+    if (!this.findStoreById(store.id) &&
+        !store.storageKey.toString().startsWith('volatile')
```
Comment (shans) [inline src/runtime/manifest.ts]: don't use toString().startsWith to test NG keys - look directly at the protocol instead. It does mean you need to have 2 code paths in the short term.

Comment (mmandlis) [reply src/runtime/manifest.ts]: done.

**Pull request merged**